BootloaderCommonPkg/Library/Ext23Lib/Ext2Fs.c

```diff
@@ -577,58 +577,117 @@ SearchDirectory (
 }
 
 /**
-  Read Superblock of the file.
+  Validate EXT2 Superblock
 
+  @param[in]      FsHandle      EXT file system handle.
   @param[in]      File          File for which super block needs to be read.
-  @param[in/out]  FileSystem    Fs on which super block is computed.
+  @param[out]     RExt2Fs       EXT2FS meta data to retreive.
 
-  @retval 0 if superblock compute is success
+  @retval 0 if superblock validation is success
   @retval other if error.
 **/
 RETURN_STATUS
 EFIAPI
-ReadSBlock (
-  IN      OPEN_FILE     *File,
-  IN OUT  M_EXT2FS      *FileSystem
+Ext2SbValidate (
+  IN CONST EFI_HANDLE  FsHandle,
+  IN CONST OPEN_FILE   *File     OPTIONAL,
+  OUT      EXT2FS      *RExt2Fs  OPTIONAL
   )
 {
   PEI_EXT_PRIVATE_DATA *PrivateData;
   UINT8 *Buffer;
-  EXT2FS Ext2Fs;
+  EXT2FS *Ext2Fs;
   UINT32 BufSize;
   RETURN_STATUS Rc;
   UINT32 SbOffset;
 
   Rc = 0;
   Buffer = NULL;
 
-  PrivateData = (PEI_EXT_PRIVATE_DATA*) File->FileDevData;
+  if (FsHandle == NULL) {
+    Rc = RETURN_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  PrivateData = (PEI_EXT_PRIVATE_DATA *)FsHandle;
 
   Buffer = AllocatePool ((PrivateData->BlockSize > SBSIZE) ? PrivateData->BlockSize : SBSIZE);
   if (Buffer == NULL) {
     Rc = EFI_OUT_OF_RESOURCES;
     goto Exit;
   }
 
-  Rc = DEV_STRATEGY (File->DevPtr) (File->FileDevData, F_READ,
-                                    SBOFF / PrivateData->BlockSize, PrivateData->BlockSize, Buffer, &BufSize);
+  if (File == NULL) {
+    Rc = BDevStrategy (PrivateData, F_READ,
+                       SBOFF / PrivateData->BlockSize, PrivateData->BlockSize, Buffer, &BufSize);
+  } else {
+    Rc = DEV_STRATEGY (File->DevPtr) (PrivateData, F_READ,
+                                      SBOFF / PrivateData->BlockSize, PrivateData->BlockSize,
+                                      Buffer, &BufSize);
+  }
+
   if (Rc != 0) {
     goto Exit;
   }
+
   SbOffset = (SBOFF < PrivateData->BlockSize) ? SBOFF : 0;
-  E2FS_SBLOAD ((VOID *)(&Buffer[SbOffset]), &Ext2Fs);
-  if (Ext2Fs.Ext2FsMagic != E2FS_MAGIC) {
-    Rc = EFI_INVALID_PARAMETER;
+  Ext2Fs = (EXT2FS *)(&Buffer[SbOffset]);
+  if (Ext2Fs->Ext2FsMagic != E2FS_MAGIC) {
+    DEBUG ((DEBUG_ERROR, "Unsupported FS magic: 0x%x\n", Ext2Fs->Ext2FsMagic));

```

Comment:
Do we consider it as an error condition ?   Since it is first time to check a FS,  it might be normal to see incorrect Magic.  For example, if the FS is FAT.   Maybe we should remove this DEBUG statement ?

Reply:
good suggestion, will update accordingly